snap/get_version

```diff
@@ -0,0 +1,116 @@
+#    b. <last tag name on current branch (not in main)>+<commit_sha>.<last_commit_merged_from_main> for other branches.
+#
+# Any of those version will be annoted with +dirty if there are local changes.
+
+# set_version will markup the version in the snapcraft.yaml file after amending it with a dirty markup if necessary.
+# $1: version: the version to set.
+set_version() {
+    version="${1}"
+
+    version=$(annotate_with_dirty "${version}")
+    craftctl set version="${version}"
+}
+
+# annotate_with_dirty may amend the version with a dirty markup if there are local changes.
+# $1: version: the version to annotate.
+annotate_with_dirty() {
+    version="${1}"
+
+    # check if current tree content is dirty.
+    is_dirty=$(git -C "${SNAPCRAFT_PART_SRC}" status --porcelain)
+    if [ -n "${is_dirty}" ]; then
+        version="${version}+dirty"
+    fi
+
+    echo "${version}"
+}
+
+# last_tag_on_branch returns the last tag on that branch that is not in another provided branch.
+# $1: origin_branch: branch we start listing from.
+# $2: other_branch: the branch where we don’t want to tags to be reachable from.
+# If we don’t have another branch to compare with, it will just return the most recent tag.
+last_tag_on_branch() {
+    origin_branch="${1}"
+    other_branch="${2}"
+
+    all_tags_on_origin_branch=$(git -C "${SNAPCRAFT_PART_SRC}" tag --sort=-committerdate --merged "${origin_branch}")
+    for tag in ${all_tags_on_origin_branch}; do
+        # Just return first commit if there is no branch to compare against the main branch.
+        if [ -z "${other_branch}" ]; then
+            tag=$(strip_branch_tag_prefix "${tag}" "${origin_branch}")
+            echo "${tag}"
+            return
+        fi
+
+        # Skip the tags that are reachable from the other branch.
+        commit_from_tag=$(git -C "${SNAPCRAFT_PART_SRC}" rev-list -n 1 "${tag}")
+        if git -C "${SNAPCRAFT_PART_SRC}" merge-base --is-ancestor "${commit_from_tag}" "${other_branch}"; then
+            continue
+        fi
+
+        tag=$(strip_branch_tag_prefix "${tag}" "${origin_branch}")
+        echo "${tag}"
+        return
+    done
+
+    echo "notag"
+}
+
+# strip_branch_tag_prefix will remove the branch name prefix from the tag name.
+# $1: tag: the tag name to strip the prefix from.
+# $2: current_branch: the branch name to strip from the tag.
+strip_branch_tag_prefix() {
+    tag="${1}"
+    current_branch="${2}"
+
+    echo "${tag#${current_branch}-}"
+}
+
+
+current_branch=$(git -C "${SNAPCRAFT_PART_SRC}" branch --show-current)
+
+# If current commit on that branch is tagged, this is the version we are using.
+current_commit_tag=$(git -C "${SNAPCRAFT_PART_SRC}" describe --exact-match --abbrev=0 --tags 2>/dev/null) || true
+if [ -n "${current_commit_tag}" ]; then
+    current_commit_tag=$(strip_branch_tag_prefix "${current_commit_tag}" "${current_branch}")
+    set_version "${current_commit_tag}"
+    exit 0
+fi
```

Comment:
Why do we not skip the current commit if it comes from main?

Reply:
This is not what the code is doing here. Here, if the current commit is tagged, then, we are using directly that tag name version. See the comment too.

Reply:
Yes, I understand what the code is doing. If the current commit is tagged, we use that tag. My question is: Why do we not do the same check as below, i.e. that the commit is from main.

Reply:
I’m still unsure to understand you, this is independent of branch? Meaning that if this is the main branch or any other and the current commit is tagged, we take this. Why would we skip the current commit if it comes from main?

Reply:
We do. As documented in the comment above it, the `last_tag_on_branch` function ignores tags which are reachable from the other branch (main).

If I add a new tag `0.2` on main and merge main into the `msentraid` branch and run the script, I get `0.1+c13d899.e6aac67`. So we do skip the `0.2` tag which comes from main.

Reply:
> This is expected, we are considering only the releasing on the current branch not coming from other.
> 
> You rejected the whole idea of "only tagging main" I had at first and not tagging any other branches. This is why then, when you are on msentraid, the tag on main is ignored.

I understand why you did that.

My question remains, why do we don't do the same for the current commit? We could do it like this instead:

```
# Get most recent tag on that branch not coming from the other branch.
tag=$(last_tag_on_branch "${current_branch}" "${other_branch}")

if [ -n "$tag" ] && [ "$tag" = "$(git describe --tags --exact-match)" ]; then
    # The tag is on the current commit, so we don't need to append the commit SHA.
    set_version "${tag}"
    exit 0
fi
```

Reply:
I don’t see the advantage of that approach? If there is a tag, we don’t append the commit SHA anyway and do an early return, as you did here?

I think we really don’t understand each other and I think we should HO before the meeting, now?

Reply:
I had to adapt this a little bit as `git describe --tags --exact-match` is failing if there is no tag on it. I think it’s fine just sending stderr to /dev/null and then, we have thus an empty tag which is an acceptable output in that situation.

Reply:
> Yes, that's why I added the [ -n "$tag" ] check before, so that we don't end up setting an empty version if the output of git describe --tags --exact-match is empty.

Right, I was rather explaining on adding `2>/dev/null` to avoid stderr output in the "expected last tag is not versionned" flow.